### DISCUSSION TITLE
Attribute fields in template

### DISCUSSION THREAD

**Discussion by jonsmith1234** [freeplane/freeplane#1767] [Help] [answered]
Is there a way to have blank attribute fields in a template (or formulas based upon the parent), so the user sees them and will fill them out?  I am assuming that a template is the way to go, but if I am wrong, please enlighten me.

I can get my formulas to work, just want a way that the mind map has the fields there automatically without the user having to add them on each node.

**Comment by euu2021** [comment]
See #959 

Maybe, you want it to be even more automatic. If so, we can think in a solution. For example, using listeners.

**Reply by jonsmith1234** [reply]
How would I go about setting that up?

My use case is:

1. Every node will have the same 3 attribute fields
2. When you create a new node, the 3 attributes are there, but blank for the user to fill in
3. One of the attributes has a formula, based upon the parent.  I've worked out the formulas, just need to embed them in the template.

**Reply by euu2021** [reply, marked as answer]
This script adds the attributes defined in the script, as soon as the node is created:
![java_6bnqJ3a8Mw](https://github.com/freeplane/freeplane/assets/77707706/3ad5de06-8418-4087-af4e-af6f86ff2aeb)

Script:

```
import org.freeplane.api.NodeChangeListener
import org.freeplane.api.NodeChanged
import org.freeplane.api.NodeChanged.ChangedElement

class MyNodeChangeListener implements NodeChangeListener {
    public void nodeChanged(NodeChanged event) {
        def createdSince = new Date()
        createdSince.setSeconds(createdSince.getSeconds() - 2)
        /* enum ChangedElement {TEXT, DETAILS, NOTE, ICON, ATTRIBUTE, FORMULA_RESULT, UNKNOWN} */
        if (event.changedElement == NodeChanged.ChangedElement.UNKNOWN) {
            if (event.node.CreatedAt.after(createdSince)) {
                if (event.node.attributes.isEmpty()) {
                    event.node.attributes.add("aaa", "xxx")
                    event.node.attributes.add("bbb", "=node.text")
                }
            }
        }
    }
}

mindMap.addListener(new MyNodeChangeListener())
```

You can choose your attributes by changing the lines (or adding more lines like these):
```
                    event.node.attributes.add("aaa", "xxx")
                    event.node.attributes.add("bbb", "=node.text")
```

Learn how to apply scripts at [this page](https://docs.freeplane.org/scripting/Start_Here.html) of the documentation.

You only need to execute the script once. But, if you restart Freeplane, the script needs to be executed again. If you want that it happens automatically, place the next script at this init folder, as explained [in the documentation](https://docs.freeplane.org/scripting/Init_scripts.html):

```
import org.freeplane.api.NodeChangeListener
import org.freeplane.api.NodeChanged
import org.freeplane.api.NodeChanged.ChangedElement

class MyNodeChangeListener implements NodeChangeListener {
    public void nodeChanged(NodeChanged event) {
        def createdSince = new Date()
        createdSince.setSeconds(createdSince.getSeconds() - 2)
        /* enum ChangedElement {TEXT, DETAILS, NOTE, ICON, ATTRIBUTE, FORMULA_RESULT, UNKNOWN} */
        if (event.changedElement == NodeChanged.ChangedElement.UNKNOWN) {
            if (event.node.CreatedAt.after(createdSince)) {
                if (event.node.attributes.isEmpty()) {
                    event.node.attributes.add("aaa", "xxx")
                    event.node.attributes.add("bbb", "=node.text")
                }
            }
        }
    }
}

c.openMindMaps.each{it.addListener(new MyNodeChangeListener())}
```
This last one will apply the listener to every map that is opened during the FP startup. Tell me if you need a different behavior.

If you need to have the script automatically executed on every new map that you create, we can think of a solution for that.

**Reply by jonsmith1234** [reply]
Fabulous, thanks!

**Comment by macmarrum** [comment]
> If you need to have the script automatically executed on every new map that you create, we can think of a solution for that.

To have your listener run automatically whenever you open a particular mind map (or create a mind map from a particular template), you can use the add-on "Script Execution On Mindmap Open"

- https://github.com/freeplane/freeplane/discussions/688

**Comment by dpolivaev** [comment]
You can assign any attributes and attribute values to styles in the style map and then use Edit->Node properties->Attributes from style to apply them to any node
